compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp

```diff
@@ -0,0 +1,674 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ * Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/QuantizeWeightsWithGPTQPass.h"
+#include "QuantizationUtils.h"
+#include "helpers/LayerInfoMap.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/CircleNodeVisitor.h>
+#include <luci/Service/Nodes/CircleConst.h>
+#include <luci/Log.h>
+#include <loco/IR/TensorShape.h>
+
+#include <iostream>
+#include <cmath>
+#include <functional>
+#include <limits>
+
+namespace
+{
+
+using namespace luci;
+using IterFunc = std::function<void(uint32_t *, loco::TensorShape &, int32_t)>;
+
+void iterate_per_channel_with_order(CircleConst *node, IterFunc func, bool reverse)
+{
+  loco::TensorShape dimension;
+  dimension.rank(4);
+  uint32_t indices[4] = {0};
+  int32_t channel_dim_index{0};
+  uint32_t num_dims[4];
+  if (!get_channel_dim_index(node, dimension, channel_dim_index))
+  {
+    throw std::runtime_error("Failed to get channel dim index.");
+  }
+
+  auto order = reverse ? std::vector<size_t>{3, 1, 2, 0} : std::vector<size_t>{0, 1, 2, 3};
+
+  for (uint32_t i = 0; i < 4; ++i)
+  {
+    num_dims[i] = dimension.dim(order[i]).value();
+  }
+
+  for (uint32_t i = 0; i < num_dims[0]; i++)
+  {
+    for (uint32_t j = 0; j < num_dims[1]; j++)
+    {
+      for (uint32_t s = 0; s < num_dims[2]; s++)
+      {
+        for (uint32_t t = 0; t < num_dims[3]; t++)
+        {
+          indices[order[0]] = i;
+          indices[order[1]] = j;
+          indices[order[2]] = s;
+          indices[order[3]] = t;
+          func(indices, dimension, channel_dim_index);
+        }
+      }
+    }
+  }
+}
+
+} // namespace
+
+namespace luci
+{
+
+namespace
+{
+
+size_t calculate_qauntized_value(CircleConst *node, uint32_t *indices, loco::TensorShape &dimension,
+                                 int channel_dim_index, std::vector<float> &scaling_factor,
+                                 std::vector<float> &max, std::vector<float> &min)
+{
+  int channel_idx = indices[channel_dim_index];
```

Comment:
add some assertion checks for input arguments

Reply:
`indices[channel_dim_index];` may produce out of bounds problem and produce segmentation fault.